### PR TITLE
W2v2 convdebug

### PR DESCRIPTION
**Pull request by taylanbil**

### CODE REVIEW

File: fairseq/criterions/wav2vec_criterion.py [hunk 1]
```diff
@@ -53,11 +63,24 @@ def forward(self, model, sample, reduce=True, log_pred=False):
...
+            loss = F.cross_entropy(
+                logits, target, reduction="sum" if reduce else "none",
+                ignore_index=-1,
+            )
         else:
-            loss = F.binary_cross_entropy_with_logits(logits, target.float(), weights, reduction="sum" if reduce else "none",)
+            loss = F.binary_cross_entropy_with_logits(
+                logits, target.float(), weights,
+                reduction="sum" if reduce else "none",
+                ignore_index=-1,
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py:74]: this wont actually work because binary cross entropy with logits does not have ignore index. you need to use reduction="none" here and then zero out the loss coming from unmasked states

File: fairseq/criterions/wav2vec_criterion.py [hunk 2]
```diff
@@ -53,11 +63,24 @@ def forward(self, model, sample, reduce=True, log_pred=False):
...
+            )
 
-        sample_size = target.numel() if self.infonce else target.long().sum().item()
+        if 'sample_size' in sample and self.infonce:
+            sample_size = sample['sample_size']
+        elif 'mask_indices' in sample['net_input'] and self.infonce:
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py:79]: maybe remove "and self.infonce" part?

File: fairseq/criterions/wav2vec_criterion.py [hunk 3]
```diff
@@ -154,4 +186,4 @@ def logging_outputs_can_be_summed() -> bool:
         across workers prior to calling `reduce_metrics`. Setting this
         to True will improves distributed training speed.
         """
-        return False
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py:157]: if you keep it at false, do you still see larger accuracy etc (i know we tried 1 node, but still)

File: fairseq/data/audio/raw_audio_dataset.py [hunk 4]
```diff
@@ -106,9 +154,131 @@ def collater(self, samples):
...
+        That's why we do it here during data prep when run on XLA.
+        """
+        self.n_negatives = self.args.num_negatives
+        self.cross_sample_negatives = self.args.cross_sample_negatives
+        if self.n_negatives == 0 and self.cross_sample_negatives == 0:
+            # FIXME: handle this
+            return y.new(0)
+
+        (bsz, tsz), fsz = mask_indices.shape, self.args.final_dim
+        high = mask_indices.sum(-1).max().item()
```
Comment (alexeib) [inline fairseq/data/audio/raw_audio_dataset.py:214]: this means that sometimes you will sample negatives from masked timesteps for examples that are shorter than the longest one. why not sample separately per each example in the batch and use correct high for each example?

File: fairseq/data/audio/raw_audio_dataset.py [hunk 5]
```diff
@@ -106,9 +154,131 @@ def collater(self, samples):
...
+                cross_neg_idxs = torch.randint(
+                    low=0,
+                    high=cross_high - 1,
+                    size=(bsz, self.cross_sample_negatives * num),
+                )
+                cross_neg_idxs[cross_neg_idxs >= tszs] += 1
+
+        if self.n_negatives > 0:
+            for i in range(1, bsz):
+                neg_idxs[i] += i * high
```
Comment (alexeib) [inline fairseq/data/audio/raw_audio_dataset.py:258]: this is problematic because previously it assumed "high" is the number of timesteps, but we've redefined high above to be something smaller. you need to do neg_idxs[i] += i * tsz here

File: fairseq/models/wav2vec/wav2vec2.py [hunk 6]
```diff
@@ -405,47 +405,66 @@ def build_model(cls, args, task=None):
...
+            #num = l // self.n_negatives if num is None else num
+            assert len(padding_counts) == bsz
+            neg_idxs = [
+                torch.randint(low=0, high=high-1, size=(1, l))
+                for pc in padding_counts
+            ]
+            neg_idxs = torch.stack(neg_idxs)
+        return neg_idxs
+
+    def sample_negatives(self, y, num, padding_counts=None):
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:467]: is this still used given that you moved this into collater??

File: fairseq/data/audio/raw_audio_dataset.py [hunk 7]
```diff
@@ -106,9 +154,131 @@ def collater(self, samples):
...
+            input["mask_channel_indices"] = mask_channel_indices
+            out['sample_size'] = mask_indices.sum().item()
+
+        out["net_input"] = input
+        return out
+
+    def _get_neg_idxs(self, mask_indices):
+        return
+
+    def _sample_negatives(self, mask_indices):
```
Comment (alexeib) [inline fairseq/data/audio/raw_audio_dataset.py:202]: where is this actually used??

File: fairseq/models/wav2vec/wav2vec2.py [hunk 8]
```diff
@@ -454,21 +473,23 @@ def sample_negatives(self, y, num):
         y = y.view(-1, fsz)  # BTC => (BxT)C
 
         cross_high = tsz * bsz
-        high = tsz
+        high = num
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:476]: num = number of actual timesteps to sample for.
tsz = number of timesteps in the batch

in your case num should equal tsz (same as negatives from everywhere + sampling extra negs that wont get used) which would already be correct since you dont shrink y. so you dont need to change this.

File: fairseq/models/wav2vec/wav2vec2.py [hunk 9]
```diff
@@ -454,21 +473,23 @@ def sample_negatives(self, y, num):
...
+        # FIXME: there's a problem here w/ tsz and num
+        # this assumes y is shrunk at this point.
         with torch.no_grad():
             assert high > 1, f"{bsz,tsz,fsz}"
 
             if self.n_negatives > 0:
                 tszs = (
-                    buffered_arange(num)
+                    buffered_arange(tsz)
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:484]: same here

File: fairseq/models/wav2vec/wav2vec2.py [hunk 10]
```diff
@@ -454,21 +473,23 @@ def sample_negatives(self, y, num):
...
                     .unsqueeze(-1)
                     .expand(-1, self.n_negatives)
                     .flatten()
                 )
 
-                neg_idxs = torch.randint(
-                    low=0, high=high - 1, size=(bsz, self.n_negatives * num)
-                )
+                neg_idxs = torch.randint(low=0, high=high-1, size=(bsz, self.n_negatives * tsz))
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:490]: here, for xla, you should loop over bsz and sample negatives individually for each example setting high to be tsz-sum(padding[b]) as we discussed. otherwise you might be sampling negatives from states that are padded. i guess padding[b] would come from padding_counts which is currently not used

File: fairseq/models/wav2vec/wav2vec2.py [hunk 11]
```diff
@@ -495,9 +516,13 @@ def sample_negatives(self, y, num):
...
             neg_idxs = torch.cat([neg_idxs, cross_neg_idxs], dim=1)
 
+        import pdb
+        pdb.set_trace()
         negs = y[neg_idxs.view(-1)]
+        import pdb
+        pdb.set_trace()
         negs = negs.view(
-            bsz, num, self.n_negatives + self.cross_sample_negatives, fsz
+            bsz, tsz, self.n_negatives + self.cross_sample_negatives, fsz
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:525]: no need to change this

File: fairseq/models/wav2vec/wav2vec2.py [hunk 12]
```diff
@@ -587,13 +627,21 @@ def forward(self, source, padding_mask=None, mask=True, features_only=False):
 
             y = self.project_q(y)
 
+            num = y.size(1) if tszs_after_mask is None else max(tszs_after_mask)
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:630]: you dont need to do this, just set it to y.size(1) always

File: fairseq/models/wav2vec/wav2vec2.py [hunk 13]
```diff
@@ -608,17 +656,24 @@ def forward(self, source, padding_mask=None, mask=True, features_only=False):
             y = self.project_q(y)
 
             if self.negatives_from_everywhere:
-                negs, _ = self.sample_negatives(unmasked_features, y.size(1))
+                negs, _ = self.sample_negatives(
+                    unmasked_features, num, padding_counts=padding_counts,
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:660]: no need to change to num here

File: fairseq/models/wav2vec/wav2vec2.py [hunk 14]
```diff
@@ -608,17 +656,24 @@ def forward(self, source, padding_mask=None, mask=True, features_only=False):
...
+                )
                 negs = self.project_q(negs)
             else:
-                negs, _ = self.sample_negatives(y, y.size(1))
+                negs, _ = self.sample_negatives(
+                    y, num,
```
Comment (alexeib) [inline fairseq/models/wav2vec/wav2vec2.py:665]: or here
